x-pack/plugins/apm/public/components/app/TransactionDetails/Transaction/WaterfallContainer/index.tsx

```diff
@@ -29,107 +31,65 @@ interface Props {
   location: any;
 }
 
-interface State {
-  services: string[];
-  hits: Array<Span | Transaction>;
+interface WaterfallRequestProps {
+  urlParams: IUrlParams;
+  transaction: Transaction;
+  render: RRRRender<WaterfallResponse>;
 }
 
-export class WaterfallContainer extends PureComponent<Props, State> {
-  public state: Readonly<State> = {
-    services: [],
-    hits: []
-  };
-
-  public maybeFetchWaterfall = async (prevProps?: Props) => {
-    const hasTrace = this.props.transaction.hasOwnProperty('trace');
-    if (hasTrace) {
-      this.maybeFetchTrace(prevProps);
-    } else {
-      this.maybeFetchSpans(prevProps);
-    }
-  };
-
-  public maybeFetchSpans = async (prevProps = {} as Partial<Props>) => {
-    const { start, end } = this.props.urlParams;
-    const { start: prevStart, end: prevEnd } =
-      prevProps.urlParams || ({} as IUrlParams);
-
-    const serviceName: string = get(this.props.transaction, SERVICE_NAME);
-    const prevServiceName: string = get(prevProps.transaction, SERVICE_NAME);
-
-    const transactionId: string = get(this.props.transaction, TRANSACTION_ID);
-    const prevTransactionId: string = get(
-      prevProps.transaction,
-      TRANSACTION_ID
+function WaterfallRequest({
+  urlParams,
+  transaction,
+  render
+}: WaterfallRequestProps) {
+  const hasTrace = transaction.hasOwnProperty('trace');
+  if (hasTrace) {
+    return (
+      <WaterfallV2Request
+        urlParams={urlParams}
+        transaction={transaction}
+        render={render}
+      />
     );
-
-    if (
-      serviceName !== prevServiceName ||
-      transactionId !== prevTransactionId ||
-      start !== prevStart ||
-      end !== prevEnd
-    ) {
-      const spans = await loadSpans({
-        serviceName,
-        transactionId,
-        start,
-        end
-      });
-
-      this.setState({
-        hits: [this.props.transaction, ...spans],
-        services: [serviceName]
-      });
-    }
-  };
-
-  public maybeFetchTrace = async (prevProps = {} as Partial<Props>) => {
-    const { start, end } = this.props.urlParams;
-    const { start: prevStart, end: prevEnd } =
-      prevProps.urlParams || ({} as IUrlParams);
-
-    const traceId = get(this.props, `transaction.${TRACE_ID}`);
-    const prevTraceId: string = get(prevProps, `transaction.${TRACE_ID}`);
-
-    if (
-      traceId &&
-      start &&
-      end &&
-      (traceId !== prevTraceId || start !== prevStart || end !== prevEnd)
-    ) {
-      const { hits, services } = await loadTrace({ traceId, start, end });
-      this.setState({ hits, services });
-    }
-  };
```

Comment:
All this was error-prone and not easy to read. It's taken care of by react-redux-request now.